### PR TITLE
enable headdims > 64 for flash attention on sm90

### PR DESCRIPTION
**Pull request by Fuzzkatt**
Follow up to #99105 which disabled FlashAttention when using autograd and mem eff attention for the following cases

head_dim > 64
sm86 or newer

We have tested enabling FlashAttention on sm90 and it works, so this PR will enable it back for sm90 and add in a test

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:623]: Did you also check that this bug is no longer showing up for mem-efficient kernel in sm90? I don't see that reflected in the tests

File: test/test_transformers.py
Comment (malfet) [inline test/test_transformers.py:1459]: Why do we need a specific test for sm_90? Wouldn't existing tests that run on sm_80 would also run on sm_90?

**Review by malfet — APPROVED** [review]
LGTM, but I wonder why we need to make a new test (existing should cover both sm_80 and sm_90, wouldn't they?)

File: test/test_transformers.py
Comment (Fuzzkatt) [reply test/test_transformers.py:1459]: From what I understand, the existing test that tests the `with sdp_kernel(enable_mem_efficient=False, enable_flash=True, enable_math=False)` path as it currently is (https://github.com/pytorch/pytorch/blob/main/test/test_transformers.py#L1415) would expect a RuntimeError for all SM86+ with headdim > 64. In particular, this would expect a RuntimeError for SM90 with headdim > 64, but this PR enables SM90 with headdim > 64 so it would work successfully, thus failing the assert. So the new test is SM90 specific and also verifies that the backward grads are numerically accurate.

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/sdp_utils.h:539]: I am landing https://github.com/pytorch/pytorch/pull/99948 which will cause merge conflicts so I would wait until this before tyring to land

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h
Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.h:627]: ??

File: test/test_transformers.py
Comment (drisspg) [reply test/test_transformers.py:1491]: You should expand the pamertrization here: https://github.com/pytorch/pytorch/blob/f274c4ecf68d989f921efb685cfb3177cf2e6220/test/test_transformers.py#L1730 
and also add a similiar failing assert like this: https://github.com/pytorch/pytorch/blob/f274c4ecf68d989f921efb685cfb3177cf2e6220/test/test_transformers.py#L1673

File: aten/src/ATen/native/transformers/cuda/sdp_utils.h
Comment (Fuzzkatt) [reply aten/src/ATen/native/transformers/cuda/sdp_utils.h]: I ran https://github.com/Fuzzkatt/pytorch/blob/Fuzzkatt/enable_flash_attention_headdims_gt64_sm90/test/test_transformers.py#L1645 which should cover this case on H100 and it works, let me know if any additional testing is needed.

File: test/test_transformers.py
Comment (Fuzzkatt) [reply test/test_transformers.py]: Thanks for the suggestion @drisspg, I updated the test_transformer.py to remove this new test and follow your suggestion to reuse the existing `test_flash_attention_vs_math_ref_grads` with added parameterization and assert for sm86 and sm89.

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:1842]: hmmm this isn't quite right, we want the forward to fail not the backward

Comment (drisspg) [reply test/test_transformers.py:1842]: unless the inputs don't require _grad then the forward can succede but this is up for debate as well

**Review by drisspg — APPROVED** [review]

File: test/test_transformers.py
Comment (Fuzzkatt) [reply test/test_transformers.py:1842]: Followed up offline. The highlighted test calls the private function directly so this is fine; the sdpa dispatcher test where we want it to fail in the forward is here: https://github.com/pytorch/pytorch/blob/97afbcbc8007857a51c85e9c61fe6d80564ef1f9/test/test_transformers.py#L1071

**Pull request closed without merging**
